libwild/src/aarch64.rs

```diff
@@ -753,6 +768,33 @@ impl crate::arch::Arch for AArch64 {
     fn rel_type_to_string(r_type: u32) -> std::borrow::Cow<'static, str> {
         aarch64_rel_type_to_string(r_type)
     }
+
+    fn write_plt_entry(
+        plt_entry: &mut [u8],
+        got_address: u64,
+        plt_address: u64,
+    ) -> crate::error::Result {
+        plt_entry.copy_from_slice(PLT_ENTRY_TEMPLATE);
+        let plt_page_address = plt_address & DEFAULT_AARCH64_PAGE_IGNORED_MASK;
+        let offset = got_address.wrapping_sub(plt_page_address) as i64;
+        anyhow::ensure!(
+            ((-1 << 32)..=1 << 32).contains(&offset),
+            "PLT is more than 4GiB away from GOT"
+        );
+        RelocationInstruction::Adr.write_to_value(
+            // The immediation value represents a distance in pages.
+            offset as u64 / DEFAULT_AARCH64_PAGE_SIZE,
```

Comment:
Division of a negative value cast to an unsigned value will give a very different result than division of the original negative value. This makes me think that perhaps this won't work if `offset` is negative. Using a bitmask instead of division might be more likely to produce the desired result. If we're not sure we want to handle negative offsets now, then you could change the ensure above to disallow negative offsets. That would mean that we wouldn't support placing the PLT after the GOT, but until such time as we support linker scripts, that's unlikely to be an issue.

Reply:
A very good point. I tend to adjust the code so that we will assume the GOT entry follows PLT entry.